### PR TITLE
Updated and fixed the api/build.py and cli/main.py files

### PR DESCRIPTION
**Pull request by ADRIDEV2024**
Hello 👋. I have been reading the responses to the changes I added and have taken them into account with the updates I am going to describe below, as I believe they could significantly improve the performance, efficiency and readability of the tool. I hope these small changes seem appropriate to you, let me know when you have time.

### **FIXS:** 

- Replaced print sentences with logging
- Updating error messages for a better understanding and user-friendly
- Added a new argument "--version" in the parse_args function (main.py). This will show the current version of the wheel
- Replaced os module with Path in wheel_file_path variable.  **Reason:**  Using os.path is more operating system error-prone and less readable than pathlib.Path
- Removed ThreadPoolExecutor and added instead ProcessPoolExecutor with a worker_count option and os.cpu_count. 
**Reason**: Currently, construction is parallelized, but the number of workers is not adequately controlled. 
**My solution:** Use concurrent.futures.ProcessPoolExecutor instead of ThreadPoolExecutor to take advantage of multiple CPUs. Use os.cpu_count() function as default for worker_count.

**NOTE**: On previous PR you said that I dropped accidentaly an important line of code, but I review that and that line was only placed to another line below, I don´t know exactly if this will make a problem in future, but it should not have to happen anything wrong.

**BENEFITS**:

- Better understanding about errors and possible bugs
- Better processing performance
- More efficiency 
- More reliability

**SCREENSHOTS:** 

code_fixed(5)
code_fixed(1)
code_fixed(2)
code_fixed

### CODE REVIEW

File: binary_wheel_builder/api/build.py [hunk 1]
```diff
@@ -14,6 +14,8 @@
 from binary_wheel_builder.wheel.reproducible import ReproducibleWheelFile
 from binary_wheel_builder.wheel.util import generate_metadata_file, generate_wheel_file
 
+logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
```
Comment (timo-reymann) [inline binary_wheel_builder/api/build.py:17]: chore: Please move this to `cli/main.py`

Logging configuration should always happen at the consumer side of things. If this is used as library code for example, this would potentially overwrite logging configuration at runtime, potentially breaking the project configuration.

File: binary_wheel_builder/api/build.py [hunk 2]
```diff
@@ -22,7 +24,7 @@ def _write_wheel(
         metadata: dict,
         wheel_file_entries: list[WheelFileEntry]
 ):
-    wheel_file_path = os.path.join(out_dir, wheel.wheel_filename(tag))
+    wheel_file_path = Path(out_dir) / wheel.wheel_filename(tag))
```
Comment (timo-reymann) [inline binary_wheel_builder/api/build.py:27]: chore: Please add the import for `pathlib.Path` 

Like the object-oriented Path api more, don't recall why I did not use it here, but dont see a reason not to.

File: binary_wheel_builder/api/build.py [hunk 3]
```diff
@@ -106,30 +108,43 @@ def build_wheel(wheel_meta: Wheel, dist_folder: Path, worker_count: int = 1) ->
     :return: Yields for each generated platform wheel
     """
     dist_folder.mkdir(exist_ok=True)
-    with concurrent.futures.ThreadPoolExecutor(max_workers=worker_count) as executor:
+  
+    worker_count = worker_count or os.cpu_count()
```
Comment (timo-reymann) [inline binary_wheel_builder/api/build.py:112]: note: The worker count was chosen deliberately

This is intended to run mostly in CI environments with [cgroups ](https://blog.nginx.org/blog/what-are-namespaces-cgroups-how-do-they-work) defined, which leads to false postives of CPU counts. To avoid weird OOM kill bugs, I set the worker count to a reasonable default with freedom to pick a custom limit.

File: binary_wheel_builder/api/build.py [hunk 4]
```diff
@@ -106,30 +108,43 @@ def build_wheel(wheel_meta: Wheel, dist_folder: Path, worker_count: int = 1) ->
...
             )
-            for python_platform in wheel_meta.platforms
-        ]
-        for future in concurrent.futures.as_completed(futures):
-            yield future.result()
-
-
+            for future in concurrent.futures.as_completed(futures):
+                if future.exception() is not None:
+                   logger.error(f"Sorry, a problem has occurred :(. Ensure all data is correct
```
Comment (timo-reymann) [inline binary_wheel_builder/api/build.py:123]: chore: Fix python syntax, multiline strings cannot be simply continued on the next line

Comment (timo-reymann) [inline binary_wheel_builder/api/build.py:123]: chore: Please avoid using emojis in error messages and log outputs

File: binary_wheel_builder/api/build.py [hunk 5]
```diff
@@ -106,30 +108,43 @@ def build_wheel(wheel_meta: Wheel, dist_folder: Path, worker_count: int = 1) ->
...
-
+            for future in concurrent.futures.as_completed(futures):
+                if future.exception() is not None:
+                   logger.error(f"Sorry, a problem has occurred :(. Ensure all data is correct 
+                   or configured. Exception: {future.exception()}", exc_info=True)
+                else:
+                    try:
+                        yield future.result()
+                    except Exception as e:
+                       logger.error(f"Unexpected result has ocurred", exc_info=True)
```
Comment (timo-reymann) [inline binary_wheel_builder/api/build.py:129]: nit: Rephrase this to `Unexpected error`  which is more well known to users unfamiliar with python and, in general, more well established

File: binary_wheel_builder/api/build.py [hunk 6]
```diff
@@ -106,30 +108,43 @@ def build_wheel(wheel_meta: Wheel, dist_folder: Path, worker_count: int = 1) ->
...
+            return WheelPlatformBuildResult(
+                checksum=hashlib.sha256(wheel.read()).hexdigest(),
+                file_path=wheel_path,
+            )
+  except (OSError, IOError) as e:
+        logger.error(f"File operation failed for platform {python_platform}: {e}", exc_info=True)
+        raise RuntimeError(f"File operation failed for platform {python_platform}: {e}")
+  except Exception as e:
+        logger.error(f"Unhandled exception in _build_wheel_for_platform for platform {python_platform}: {e}", exc_info=True)
+        raise
```
Comment (timo-reymann) [inline binary_wheel_builder/api/build.py:150]: chore: Please avoid logging directly in lib functionality

This makes it easier to have a central place for logging and allow consumers to chose their own level of verbosity and error handling.

File: binary_wheel_builder/cli/main.py [hunk 7]
```diff
@@ -25,6 +25,11 @@ def _parse_args(args) -> Namespace:
         type=int,
         help="Number of parallel workers to use at most for building wheels"
     )
+    parser.add_argument(
+        '--version',
+        action='version',
+        version='Wheel Builder 1.0.0'
```
Comment (timo-reymann) [inline binary_wheel_builder/cli/main.py:31]: chore: Please remove this hardcoded version number here (and potentially the entire version parameter)

Currently the version of the application is not stored to any file or available at runtime apart from using `importlib` for this, which tends to be rather unstable.

I would omit this one from the PR and make that change later on, with a proper version being available at runtime.

**Review by timo-reymann — CHANGES_REQUESTED** [review]

**Comment by ADRIDEV2024** [comment]
> Heya @ADRIDEV2024,
> 
> please take the following as a friendly advice and not as criticism. As a disclaimer, I tend to write rather directly, which is not meant as a personal attack.
> 
> I have a few more general notes on your pull request(s). Could it be that this is the first time you contributed via GitHub Pull requests to another project?
> 
> Thanks for taking the time to check out the tool and come up with some good enhancements. But especially that you took the time to create an PR. A lot of open-source stuff is happening from a few people, and it's always difficult to get into.
> 
> Please don't attach screenshots of code into PRs. The context is provided by the diff and the description, and one can always check out code at a given revision or even check the file tree using the Web UI. They simply bloat up the PR without real value, while being an additional effort for you.
> 
> While the little introduction section in the PR might look like a good idea, it does not add any additional value and looks very GenAI generated to me. This degrades the first impression of an PR significantly, and, depending on the project, it possibly even get ignored entirely.
> 
> Pull requests don't need to be perfect from the very beginning. That's why there is a review process attached, and the idea is generally more a platform to also discuss and refine the code. So please don't create a new PR for changes, but instead let's work on one until we are both happy with it and your contribution can be merged to main.
> 
> For the description, sometimes less is more, here you are duplicating plenty of things and again, they look very generated to me.
> 
> **Start Picky preferences**
> 
> A nitpicky thing from my side, I would mainly create branches, even in forks. This makes it easier for you to keep up with upstreams and allows you to work on different fixes / ideas like you would in your projects. While for this rather small project it is not too relevant, it is a good practice to have for other contexts. Introducing this habit early on will save you a few headaches, take it from someone who also started out like this.
> 
> While the commit messages of this PR don't matter too much as I will most likely squash them anyway, it is always a good idea to read the contribution guidelines where a specific style would be mentioned. That's also definitely on me in that case I didn't add it properly, but the project uses conventional commits. To fit in better, it is sometimes good if uncertain to simply check the git history, which makes this obvious. This is rather picky from me, and you might as well ignore this section, though.
> 
> **End picky preferences**
> 
> Again, thanks for your contribution. I will take a look at this PR, and I am confident we will get your great ideas merged 👍


Hi, I really appreciate those tips as I think they will be useful in my future PRs. I am not very fluent yet in creating PRs, I admit it, but anyway I try to contribute in open-source projects as much as possible and I think I put enough effort in it.
I hope that soon we will be able to merge some changes and the project will move forward, that's all I want. Thanks for those tips, I'll keep them in mind next time 👍.

File: binary_wheel_builder/cli/main.py (hunk 7 above)
Comment (ADRIDEV2024) [reply binary_wheel_builder/cli/main.py:31]: Ok, maybe it´s not the right moment to add this argument, I will remove it, I was thinking about it and is not necessary at all, but thanks for talk me about that.

File: binary_wheel_builder/api/build.py (hunk 6 above)
Comment (ADRIDEV2024) [reply binary_wheel_builder/api/build.py:150]: Ok, I will try to fix that. What do you suggest me to do instead?

File: binary_wheel_builder/api/build.py (hunk 4 above)
Comment (ADRIDEV2024) [reply binary_wheel_builder/api/build.py:123]: Oops sorry, It´s a bad habit I have sometimes when I write error messages in my projects 😅. Won´t happen again

Comment (ADRIDEV2024) [reply binary_wheel_builder/api/build.py:123]: Ok, thanks

File: binary_wheel_builder/api/build.py (hunk 3 above)
Comment (ADRIDEV2024) [reply binary_wheel_builder/api/build.py:112]: Good. I didn´t think about that possible bugs

File: binary_wheel_builder/api/build.py (hunk 6 above)
Comment (timo-reymann) [reply binary_wheel_builder/api/build.py:150]: I see two options in this case:

- raise exception as is
- add base exception and rethrow it

for the second option sth like

```python
WheelBuildException(Exception):
  pass

except Exception as e:
  raise WheelBuildException("Unhandled exception in _build_wheel_for_platform for platform ...") from e
```

File: binary_wheel_builder/cli/main.py (hunk 7 above)
Comment (timo-reymann) [reply binary_wheel_builder/cli/main.py:31]: But good idea 👍 - will definitely add this

**Comment by timo-reymann** [comment]
There are still a few minor python errors, patched them locally:

```diff
---
Index: binary_wheel_builder/cli/main.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/binary_wheel_builder/cli/main.py b/binary_wheel_builder/cli/main.py
--- a/binary_wheel_builder/cli/main.py	(revision f8850af20bac2f54345b8cd97d68c4167c9cfa7f)
+++ b/binary_wheel_builder/cli/main.py	(date 1734078216106)
@@ -1,12 +1,14 @@
+import logging
 import sys
 from argparse import ArgumentParser, Namespace
 from pathlib import Path
-import logging
+
 from binary_wheel_builder.api import build_wheel
 
 logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
 logger = logging.getLogger(__name__)
 
+
 def _parse_args(args) -> Namespace:
     parser = ArgumentParser("CLI Wheel Builder")
     parser.add_argument(
@@ -38,18 +40,17 @@
 
     if argv is None:
         argv = sys.argv[1:] if sys.argv else []
-
     args = _parse_args(argv)
 
     dist_path = Path(args.dist_folder)
-    
-   try:
+
+    try:
         dist_path.mkdir(exist_ok=True)
     except OSError as e:
         raise SystemExit(f"Failed to create dist folder at '{dist_path}': {e}")
 
     from binary_wheel_builder.cli.config_file import load_wheel_spec_from_yaml
-     
+
     try:
         wheel = load_wheel_spec_from_yaml(Path(args.wheel_spec))
     except Exception as e:
@@ -59,4 +60,4 @@
         for result in build_wheel(wheel, dist_path, worker_count=args.max_workers):
             print(f"> {result.checksum} - {result.file_path}")
     except Exception as e:
-          raise SystemExit(f"Error occurred while building wheels: {e}")
+        raise SystemExit(f"Error occurred while building wheels: {e}")
Index: binary_wheel_builder/api/build.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/binary_wheel_builder/api/build.py b/binary_wheel_builder/api/build.py
--- a/binary_wheel_builder/api/build.py	(revision f8850af20bac2f54345b8cd97d68c4167c9cfa7f)
+++ b/binary_wheel_builder/api/build.py	(date 1734078216100)
@@ -4,11 +4,11 @@
 import concurrent.futures
 import hashlib
 import os
+import pathlib
 from collections.abc import Generator
 from operator import attrgetter
-from zipfile import Path
 from pathlib import Path
-from logging import logger
+
 from binary_wheel_builder import wrapper_templates
 from binary_wheel_builder.api.meta import (Wheel, WheelFileEntry, WheelPlatformBuildResult, WheelPlatformIdentifier,
                                            WheelSource)
@@ -16,6 +16,10 @@
 from binary_wheel_builder.wheel.util import generate_metadata_file, generate_wheel_file
 
 
+class WheelBuildException(Exception):
+    pass
+
+
 def _write_wheel(
         out_dir: str,
         wheel: Wheel,
@@ -23,7 +27,7 @@
         metadata: dict,
         wheel_file_entries: list[WheelFileEntry]
 ):
-    wheel_file_path = pathlib.Path(out_dir) / wheel.wheel_filename(tag))
+    wheel_file_path = pathlib.Path(out_dir) / wheel.wheel_filename(tag)
 
     entries = [
         *wheel_file_entries,
@@ -95,7 +99,8 @@
     )
 
 
-def build_wheel(wheel_meta: Wheel, dist_folder: Path, worker_count: int = 1) -> Generator[WheelPlatformBuildResult, None, None]:
+def build_wheel(wheel_meta: Wheel, dist_folder: Path, worker_count: int = 1) -> Generator[
+    WheelPlatformBuildResult, None, None]:
     """
     Build a given wheel based on metadata and write all wheels to the dist folder.
 
@@ -107,7 +112,7 @@
     :return: Yields for each generated platform wheel
     """
     dist_folder.mkdir(exist_ok=True)
-  
+
     worker_count = worker_count or os.cpu_count()
     with concurrent.futures.ProcessPoolExecutor(max_workers=worker_count) as executor:
         futures = [
@@ -117,20 +122,24 @@
                 python_platform,
                 wheel_meta
             )
-            for future in concurrent.futures.as_completed(futures):
-                if future.exception() is not None:
-                   logger.error(f"Sorry, a problem has occurred. Ensure all data is correct 
-                                or configured. Exception: {future.exception()}", exc_info=True)
-                else:
-                    try:
-                        yield future.result()
-                    except Exception as e:
-                       logger.error(f"Unexpected error has ocurred", exc_info=True)
-                       
+            for python_platform in wheel_meta.platforms
+        ]
+
+        for future in concurrent.futures.as_completed(futures):
+            if future.exception() is not None:
+                raise WheelBuildException(
+                    "Unexpected error has occurred. Ensure all data is correct or configured"
+                ) from future.exception()
+
+            try:
+                yield future.result()
+            except Exception as e:
+                raise WheelBuildException("Unexpected error has occurred") from e
+
 
 def _build_wheel_for_platform(dist_folder, python_platform, wheel_meta):
-  try: 
-      wheel_path = _write_platform_wheel_with_wrappers(
+    try:
+        wheel_path = _write_platform_wheel_with_wrappers(
             dist_folder.__str__(),
             wheel_meta,
             python_platform,
@@ -141,12 +150,7 @@
                 checksum=hashlib.sha256(wheel.read()).hexdigest(),
                 file_path=wheel_path,
             )
-  except (OSError, IOError) as e:
-        logger.error(f"File operation failed for platform {python_platform}: {e}", exc_info=True)
-        raise RuntimeError(f"File operation failed for platform {python_platform}: {e}")
-
-  WheelBuildException(Exception):
-     pass
-
-  except Exception as e:
-      raise WheelBuildException("Unhandled exception in _build_wheel_for_platform for platform ...") from e
+    except (OSError, IOError) as e:
+        raise RuntimeError(f"File operation failed for platform {python_platform}") from e
+    except Exception as e:
+        raise WheelBuildException("Unhandled exception in _build_wheel_for_platform for platform ...") from e
```

**Comment by timo-reymann** [comment]
Added version info in [3.10.0](https://github.com/timo-reymann/python-binary-wheel-builder/releases/tag/3.10.0)

**Comment by timo-reymann** [comment]
Thanks for taking your time to contribute and considering my feedback :heart:

**Comment by ADRIDEV2024** [comment]
> Thanks for taking your time to contribute and considering my feedback ❤️

Thanks you too
